src/line_detect/src/LineDetector.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Created By: Cameron Newton
+ * Created On: July 17, 2018
+ * Detect line in image
+ */
+
+#include "LineDetector.h"
+#include <iostream>
+
+LineDetector::LineDetector() {
+    cannyLow_ = 50;
+
+    houghLinesThreshold_ = 50;
+
+    houghLinesMinLength_ = 50;
+
+    houghLinesMaxLineGap_ = 50;
+}
+
+LinesToFollow LineDetector::initialize(const cv::Mat mat_in) {
+    // Blur image and apply Canny edge detector and HoughLines
+    cv::Mat medianBlurred;
+    cv::medianBlur(mat_in, medianBlurred, 7);
+    cv::Mat dst, contours;
+    cv::Canny(medianBlurred, dst, cannyLow_, cannyLow_ * 3, 3);
+
+    imagePixelWidth_  = dst.cols;
+    imagePixelHeight_ = dst.rows;
+
+    LinesToFollow linesToFollow = defaultLinesToFollow();
+
+    std::vector<cv::Vec4i> detectedLines;
+    cv::HoughLinesP(dst,
+                    detectedLines,
+                    1,
+                    CV_PI / 360,
+                    houghLinesThreshold_,
+                    houghLinesMinLength_,
+                    houghLinesMaxLineGap_);
+
+    // Detect markers in image
+    std::vector<LineToFollow> markers = findMarkers(detectedLines);
+
+    if (markers.empty()) return linesToFollow;
+
+    // Filter through lines to get marker in top half of vision
+    linesToFollow.frontLine = *std::max_element(
+    markers.begin(), markers.end(), [&](LineToFollow lhs, LineToFollow rhs) {
+        if (lhs.frontOfMarker < (imagePixelHeight_ / 2)) {
+            return (lhs.frontOfMarker > rhs.frontOfMarker);
+        }
+    });
+
+    // Filter through lines to get marker in bottom half of vision
+    linesToFollow.rearLine = *std::min_element(
+    markers.begin(), markers.end(), [&](LineToFollow lhs, LineToFollow rhs) {
+        if ((lhs.frontOfMarker > (imagePixelHeight_ / 2))) {
+            return (lhs.frontOfMarker < rhs.frontOfMarker);
+        }
+    });
+
+    return linesToFollow;
+}
+
+std::vector<LineToFollow>
+LineDetector::findMarkers(std::vector<cv::Vec4i> allDetectedLines) {
+    std::vector<cv::Vec4i>::iterator it1, it2;
+
+    std::vector<LineToFollow> markers;
+
+    it1 = allDetectedLines.begin();
+
+    for (it1; it1 != allDetectedLines.end(); it1++) {
+        for (it2 = it1 + 1; it2 != allDetectedLines.end();) {
+            // Calculate width of lines, only return lines that are between 10
+            // and 300 pixels apart
+
+            float width = calculateWidth(*it1, *it2);
+
+            if ((width < 300) && (width > 10)) {
```

Comment:
again a general comment of intention here would help future readers/debuggers

Reply:
eh it's still good to know for future readers what the goal of this code section is, I'm not really what it's doing